### ISSUE TITLE
feature request: stop rtags from reopenning already closed files

### ISSUE THREAD

**Issue by Mafeuser** [Andersbakken/rtags#1054]
use case: during working with C/C++ project1 there is a need to switch to project2. After some while one needs to forget project2 and continue work with project1, so all the project2 buffers are chosen (projectile helps here), then closed in one turn. When rtags-location-stack-back() is invoked (on and on), project2 files will be reopened, because these are on rtags-location-stack.

   There are two example solutions below for the case.

   Solution 1.

   1. add rtags-remove-file-from-rtags-location-stack hook and add it to kill-buffer-hook.
      the hook will remove all entries from rtags-location-stack that match file being closed then will update rtags-location-stack-index to keep valid and meaningful index.
   2. add new kill-emacs hook to clean rtags-location-stack. Thanks to this emacs, while exiting will not spend time on cleaning location stack, which is useless anyway.

   visualized location stack will be updated accordingly, ie all deleted buffers will disappear from it.

   Cons: every kill of any c-mode or c++-mode file, would imply traversing through rtags-location-stack (max 100 elems long by default) to remove the file from the stack.

   Solution 2.
   
   while traversing through rtags-location-stack, rtags-location-stack-back/forward commands check if file under point is opened (get-file-buffer can be used). If not, just silently passes to next location.

   In this solution, rtags-location-stack-visualize will also display just deleted buffers, thus if one explicitly chooses entry pointing to any deleted buffer, it will be re-opened.

   Do You see any other solutions?
   What are other weak points of the above?

   regards,
   Mat.

**Comment by Andersbakken** [comment]
Sorry about the delay. I think I prefer solution 1 and to leave the decision up to the user.

I added a defun rtags-location-stack-filter that you could hook up to kill-buffer-hook if you prefer.

I even added some documentation. 

In short you should be able to do something like this:

(add-hook 'kill-buffer-hook (lambda () (rtags-location-stack-filter (buffer-file-name)))

Let me know if this works for you.

**Comment by vibrys** [comment]
@Andersbakken, I've just found this commit which aims at functionality I was looking for.

Unfortunately it does not work for me.
I had to change the hook to the following
```
(add-hook 'kill-buffer-hook (lambda () (and (buffer-file-name) (rtags-location-stack-filter (buffer-file-name)))))
```

then IMO, line https://github.com/Andersbakken/rtags/blob/master/src/rtags.el#L2298 should look like following:
```
(string= path/lambda/rx (match-string-no-properties 1 location))
```

the changes above do not suffice, I mean the entries get deleted from rtags-location-stack as expected, but rtags-location-stack-index is not touched, thus its value can be larger than number entries within rtags-location-stack, which IMO is the reason.


anyway, thank You for work on this feature, I'll find it very usefull once it's back to work.
